maestro-studio/server/src/main/java/maestro/studio/MaestroStudio.kt

```diff
@@ -103,16 +107,39 @@ object MaestroStudio {
         }.start()
     }
 
-    private fun treeToElements(tree: TreeNode, list: MutableList<UIElement> = mutableListOf()): List<UIElement> {
-        tree.children.forEach { child ->
-            treeToElements(child, list)
+    private fun treeToElements(tree: TreeNode): List<UIElement> {
+        fun gatherElements(tree: TreeNode, list: MutableList<TreeNode>): List<TreeNode> {
+            tree.children.forEach { child ->
+                gatherElements(child, list)
+            }
+            list.add(tree)
+            return list
+        }
+        fun TreeNode.attribute(key: String): String? {
+            val value = attributes[key]
+            if (value.isNullOrEmpty()) return null
+            return value
+        }
```

Comment:
**Big** nit: Should this extension function be defined outside of `treeToElements` next to `TreeNode.bounds()`? Do all the defined functions in `treeToElements` get recreated on every invocation?